### PR TITLE
Verify connection && Refactor async connection framework

### PR DESCRIPTION
**Pull request by liyuheng55555**

### CODE REVIEW

File: iotdb-core/node-commons/src/main/java/org/apache/iotdb/commons/client/async/AsyncDataNodeExternalServiceClient.java
Comment (liyuheng55555) [inline iotdb-core/node-commons/src/main/java/org/apache/iotdb/commons/client/async/AsyncDataNodeExternalServiceClient.java:1]: maybe extract some code

File: iotdb-protocol/thrift-commons/src/main/thrift/common.thrift
Comment (liyuheng55555) [inline iotdb-protocol/thrift-commons/src/main/thrift/common.thrift:1]: move ?

File: iotdb-protocol/thrift-datanode/src/main/thrift/client.thrift [hunk 1]
```diff
@@ -659,4 +659,10 @@ service IClientRPCService {
   TSBackupConfigurationResp getBackupConfiguration();
 
   TSConnectionInfoResp fetchAllConnectionsInfo();
+
+  /** For other node's call */
+  common.TSStatus testConnection()
```
Comment (liyuheng55555) [inline iotdb-protocol/thrift-datanode/src/main/thrift/client.thrift:664]: rename to testConnectionEmptyRPC

**Review by liyuheng55555 — COMMENTED** [review]
still need to improve

**Review by liyuheng55555 — COMMENTED** [review]
almost done

**Review by liyuheng55555 — COMMENTED** [review]
LGTM

File: iotdb-protocol/thrift-commons/src/main/thrift/common.thrift [hunk 2]
```diff
@@ -191,6 +191,40 @@ struct TLicense {
...
+
+struct TTestConnectionResult {
+  1: required TServiceProvider serviceProvider
+  2: required TSender sender
+  3: required bool success
+  4: optional string reason
+}
+
+struct TTestConnectionResp {
+  1: optional TSStatus status
```
Comment (OneSizeFitsQuorum) [inline iotdb-protocol/thrift-commons/src/main/thrift/common.thrift:219]: check

File: iotdb-core/confignode/src/main/java/org/apache/iotdb/confignode/manager/ClusterManager.java [hunk 3]
```diff
@@ -79,4 +104,195 @@ private void generateClusterId() {
...
+              if (configNodeResp.isSetResultList()) {
+                resp.getResultList().addAll(configNodeResp.getResultList());
+              } else {
+                resp.getResultList()
+                    .addAll(
+                        badConfigNodeConnectionResult(
+                            anotherConfigNodeLocationMap.get(nodeId), nodeLocations));
+              }
+            });
+    // For DataNode
```
Comment (OneSizeFitsQuorum) [inline iotdb-core/confignode/src/main/java/org/apache/iotdb/confignode/manager/ClusterManager.java:146]: try to make dn and cn parallel

Comment (liyuheng55555) [reply iotdb-core/confignode/src/main/java/org/apache/iotdb/confignode/manager/ClusterManager.java:146]: seems not very easy

File: iotdb-core/node-commons/src/main/java/org/apache/iotdb/commons/client/request/Utils.java [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+          Function<ServiceProviderLocation, TEndPoint> getEndPoint,
+          TServiceType serviceType,
+          RequestType requestType,
+          Consumer<AsyncRequestContext<Object, TSStatus, RequestType, ServiceProviderLocation>>
+              sendRequest) {
+    // prepare request context
+    Map<Integer, ServiceProviderLocation> nodeLocationMap =
+        nodeLocations.stream().collect(Collectors.toMap(getId, location -> location));
+    AsyncRequestContext<Object, TSStatus, RequestType, ServiceProviderLocation> requestContext =
+        new AsyncRequestContext<>(requestType, new Object(), nodeLocationMap);
```
[the remark this answers is not in the document]
Comment (liyuheng55555) [reply iotdb-core/node-commons/src/main/java/org/apache/iotdb/commons/client/request/Utils.java]: cannot

File: iotdb-core/confignode/src/main/java/org/apache/iotdb/confignode/client/async/handlers/rpc/SubmitTestConnectionTaskToConfigNodeRPCHandler.java [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+  @Override
+  public void onComplete(TTestConnectionResp resp) {
+    responseMap.put(requestId, resp);
+    nodeLocationMap.remove(requestId);
+    countDownLatch.countDown();
+  }
+
+  @Override
+  public void onError(Exception e) {
+    responseMap.put(requestId, new TTestConnectionResp());
```
Comment (OneSizeFitsQuorum) [inline iotdb-core/confignode/src/main/java/org/apache/iotdb/confignode/client/async/handlers/rpc/SubmitTestConnectionTaskToConfigNodeRPCHandler.java:63]: e

File: iotdb-core/confignode/src/main/java/org/apache/iotdb/confignode/manager/ClusterManager.java (hunk 3 above)
Comment (OneSizeFitsQuorum) [reply iotdb-core/confignode/src/main/java/org/apache/iotdb/confignode/manager/ClusterManager.java:146]: add todo here

File: iotdb-core/confignode/src/main/java/org/apache/iotdb/confignode/manager/ClusterManager.java [hunk 6]
```diff
@@ -79,4 +104,191 @@ private void generateClusterId() {
...
+                resp.getResultList()
+                    .addAll(
+                        badDataNodeConnectionResult(
+                            anotherDataNodeLocationMap.get(nodeId), nodeLocations));
+              }
+            });
+    return resp;
+  }
+
+  public List<TTestConnectionResult> doConnectionTest(TNodeLocations nodeLocations) {
```
Comment (OneSizeFitsQuorum) [inline iotdb-core/confignode/src/main/java/org/apache/iotdb/confignode/manager/ClusterManager.java:176]: return TTestConnectionResp?

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/protocol/client/cn/DnToCnRequestType.java [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iotdb.db.protocol.client.cn;
+
+public enum DnToCnRequestType {
```
Comment (OneSizeFitsQuorum) [inline iotdb-core/datanode/src/main/java/org/apache/iotdb/db/protocol/client/cn/DnToCnRequestType.java:22]: add some comments

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/protocol/client/dn/DnToDnRequestType.java [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iotdb.db.protocol.client.dn;
+
+public enum DnToDnRequestType {
```
Comment (OneSizeFitsQuorum) [inline iotdb-core/datanode/src/main/java/org/apache/iotdb/db/protocol/client/dn/DnToDnRequestType.java:22]: add some comments

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/plan/execution/config/sys/TestConnectionTask.java [hunk 9]
```diff
@@ -0,0 +1,229 @@
...
+
+  private static int calculateConnectionMaxLen(TTestConnectionResp resp) {
+    return resp.getResultList().stream()
+        .map(TestConnectionTask::connectionResultToString)
+        .mapToInt(String::length)
+        .max()
+        .getAsInt();
+  }
+
+  private static String addLineBreak(String origin, int interval) {
```
Comment (OneSizeFitsQuorum) [inline iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/plan/execution/config/sys/TestConnectionTask.java:218]: clean

**Review by OneSizeFitsQuorum — COMMENTED** [review]
Add IT

**Review by OneSizeFitsQuorum — APPROVED** [review]
LGTM

**Pull request merged**